Improve the robustness of closing open file descriptors in jailer

gorbak25 opened the pull request:
## Changes
Closes all open FDs in a single syscall. In case close_range is not available then fallback to the old approach. In case the fallback fails then discover open FDs by reading from /proc/self/fds.

## Reason
Fixes #3542 - a bug where firecracker would end up spending minutes starting which arises when running on systems with OPEN_MAX set to a very high number.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license. For more information on following
Developer Certificate of Origin and signing off your commits, please check
[`CONTRIBUTING.md`][3].

## PR Checklist

- [x] If a specific issue led to this PR, this PR closes the issue.
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] API changes follow the [Runbook for Firecracker API changes][2].
- [x] User-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.
- [x] New `TODO`s link to an issue.
- [x] Commits meet [contribution quality standards](https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md#contribution-quality-standards).

---

- [x] This functionality cannot be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md
[3]: https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md

JonathanWoollett-Light commented:
Firecracker currently has full supported for 4.14 ([kernel-policy](https://github.com/firecracker-microvm/firecracker/blob/main/docs/kernel-policy.md)).

I believe the [`close_range`](https://man7.org/linux/man-pages/man2/close_range.2.html) syscall was introduced in 5.9 in https://github.com/torvalds/linux/blob/master/include/uapi/linux/close_range.h.

Please correct me if I'm wrong, but otherwise I think this was what prevented us using this approach previously.

gorbak25 commented:
@JonathanWoollett-Light I just pushed a V2 of this patch. Changes compared to V1:
- Each of the fall backs has a separate test
- If OPEN_MAX is bellow 1024 then I iterate to 1024 anyway, I decreased the maximum threshold from  2**20 to 2**16
- I replaced `std::fs::read_dir` with `nix::dir::Dir::open` in order to not close the dirfd descriptor twice - otherwise the rust code would panic. Is using the nix crate appropriate here? I added it cause it is a peer dependency anyway.

gorbak25 commented:
@JonathanWoollett-Light renamed the patch to a more appropriate name

JonathanWoollett-Light commented:
As separate Firecracker and Jailer versions are released for the 5.10 and 4.14 kernels respectively I think we can do this check at compile time.

If we include a `build.rs` in the `jailer` with:
```rust
fn main() {
    let process = std::process::Command::new("uname").arg("-r").output().unwrap();
    assert!(process.status.success());
    let version = std::str::from_utf8(&process.stdout).unwrap();
    let mut iter = version.split(".");
    let major = iter.next().unwrap().parse::<u32>().unwrap();
    let minor = iter.next().unwrap().parse::<u32>().unwrap();

    // Firecracker and the jailer officaily supports 2 kernel versions 5.10 and 4.14.
    //
    // Some improved functionality is supported in the 5.10 version that is not supported in 4.14,
    // to accmodate this we set conditionl compilation flags for these versions.
    //
    // In developing Firecracker and the jailer a newer kernel may be used, to support this we don't
    // match the versions exactly.
    match (major,minor) {
        (5.., 10..) => println!("cargo:rustc-cfg=kernel=\"5.10\""),
        (4, 14..) => println!("cargo:rustc-cfg=kernel=\"4.14\""),
        _ => unreachable!()
    }
}
```
and then change `sanitize_process` to:
```rust
fn sanitize_process() {
    // First thing to do is make sure we don't keep any inherited FDs
    // other that IN, OUT and ERR.
    #[cfg(kernel="5.10")]
    {
        // SAFETY: The values are constant and valid.
        unsafe {
            libc::syscall(
                libc::SYS_close_range,
                3,
                libc::c_uint::MAX,
                libc::CLOSE_RANGE_UNSHARE,
            )
        }
    }
    #[cfg(kernel="4.14")]
    {
        // SAFETY: Always safe.
        let fd_limit = i32::try_from(unsafe { libc::sysconf(libc::_SC_OPEN_MAX) }).unwrap();
        // Close all file descriptors excluding 0 (STDIN), 1 (STDOUT) and 2 (STDERR).
        for fd in 3..fd_limit {
            // SAFETY: Safe because close() cannot fail when passed a valid parameter.
            unsafe {
                libc::close(fd);
            }
        }
    }

    // Cleanup environment variables
    clean_env_vars();
}
```
it should fix this.

Does this solution appear good to you?

gorbak25 commented:
@JonathanWoollett-Light any feedback on what approach this PR should take?

JonathanWoollett-Light commented:
I think the best approach here is to initially address the issue with the simplest solution.

This 1st commit could implement a change like:
```rust
fn sanitize_process() {
    // First thing to do is make sure we don't keep any inherited FDs
    // other that IN, OUT and ERR.

    // Try using the close_range syscall to close all open FDs in the range of 4..UINT_MAX
    // SAFETY: The parameters are valid constants.
    let res = unsafe { libc::syscall(
        libc::SYS_close_range,
        3,
        libc::c_uint::MAX,
        libc::CLOSE_RANGE_UNSHARE,
    ) };

    // If `SYS_close_range` fails as it is not supported (as is the case on a pre 5.9 kernel), we
    // instead close by iterating through all possible file descriptors.
    match res {
        0 => (),
        // SAFETY: Always safe.
        -1 => match unsafe { *libc::__errno_location } {
            0 => (),
            // The syscall failed due to not being supported.
            libc::ENOSYS => {
                // SAFETY: Always safe.
                let fd_limit = i32::try_from(unsafe { libc::sysconf(libc::_SC_OPEN_MAX) }).unwrap();
                // Close all file descriptors excluding 0 (STDIN), 1 (STDOUT) and 2 (STDERR).
                for fd in 3..fd_limit {
                    // SAFETY: Safe because close() cannot fail when passed a valid parameter.
                    unsafe {
                        libc::close(fd);
                    }
                }
            },
            _ => panic!()
        },
        _ => unreachable!()
    }

    // Cleanup environment variables
    clean_env_vars();
}
```

I am unsure about pulling in Nix with the fallback this will need to be considered. I think using Nix to iterate through `/proc` should be a separate commit, if we want to get the 1st commit merged more quickly it should be a separate PR.

gorbak25 commented:
@JonathanWoollett-Light will come back to this PR on wednesday, I'm on a trip without my laptop right now.

dianpopa commented:
hello @gorbak25,

Thanks a lot for the proposed fix. The changes look reasonable to me in general, there are a couple of things I would like to adjust:
* firstly, I would like to drop the check that looks at the `_SC_OPEN_MAX ` variable and keep the one that calls into `close_range` and then fall back to using `proc/self/fd`. I tested a bit the performance of each method and for a system with `_SC_OPEN_MAX` 65535 I obtained around 47ms that are always spent. For kernels < 5.9, the behavior would be exactly the same as it was before the regression.
* also I would like to improve a bit on the error propagation of the code

Since this issue introduces a customer-facing regression, I would like to speed up the merging process so I opened a [separate PR](https://github.com/firecracker-microvm/firecracker/pull/3595) that starts from your commit and contains the proposed adjustments. Your contribution is mentioned there and kept you as the commit author! Please take a look at this alternate PR and review.

Thanks a lot again for looking into this!

dianpopa commented:
Closing in favor of https://github.com/firecracker-microvm/firecracker/pull/3595. The fix will be present in v1.4.0 (our next release).
Thanks @gorbak25 again for the contribution!
